### ISSUE TITLE
Missing Error for Modules being emitted in strict mode

### ISSUE THREAD

**Issue by kitsonk** [Microsoft/TypeScript#7339]
Related to #3714.

With the changes to emitting modules with the `"use strict";` prolog, TypeScript is not applying the appropriate level of code checking to catch block scoped function declaration.

**TypeScript Version:** 

1.8.2+

**Code**

[Playground Link](http://www.typescriptlang.org/Playground#src=export%20function%20foo()%20%7B%0A%09return%20'foo'%3B%0A%7D%0A%0Aif%20(true)%20%7B%0A%09function%20bar()%20%7B%20%2F*%20should%20throw%20error%20that%20functions%20must%20be%20declared%20at%20the%20top%20level%20*%2F%0A%09%09return%20'bar'%3B%0A%09%7D%0A%7D%0A)

```ts
export function foo() {
	return 'foo';
}

if (true) {
	function bar() { /* should throw error that functions must be declared at the top level */
		return 'bar';
	}
}
```

**Expected behavior:**

Error at block level function declaration.

**Actual behavior:** 

No error and an invalid emit.

**Comment by kitsonk** [comment]
Just to clarify, the above is when targeting ES5, as block scoped functions in strict mode are allowed, but TypeScript isn't erroring even [on this](http://www.typescriptlang.org/Playground#src=export%20function%20foo()%20%7B%0A%09return%20'foo'%3B%0A%7D%0A%0Aif%20(false)%20%7B%0A%09function%20bar()%20%7B%20%2F*%20should%20throw%20error%20that%20functions%20must%20be%20declared%20at%20the%20top%20level%20*%2F%0A%09%09return%20'bar'%3B%0A%09%7D%0A%7D%0A%0Abar()%3B%20%2F*%20should%20even%20complain%20about%20this%20if%20assuming%20ES6%20*%2F%0A):

```typescript
export function foo() {
	return 'foo';
}

if (false) {
	function bar() { /* should throw error that functions must be declared at the top level */
		return 'bar';
	}
}

bar(); /* should also complain about this, if it was assuming ES6 */
```

**Comment by vladima** [comment]
duplicate of #3714

**Comment by kitsonk** [comment]
@vladima are you sure?  I agree it is related, but it is not a dupe.

That is about rewriting duplicate function names.  Currently TypeScript isn't throwing in strict mode when referencing a block scoped function.  TypeScript thinks there is nothing wrong [with this](http://www.typescriptlang.org/Playground#src='use%20strict'%3B%0A%0Aif%20(true)%20%7B%0A%20%20function%20bar()%20%7B%0A%20%20%20%20console.log('bar')%3B%0A%20%20%7D%0A%7D%0A%0Abar()%3B%0A):

```typescript
'use strict';

if (true) {
  function bar() {
    console.log('bar');
  }
}

bar();
```

But no matter what, it throws an error at runtime... In those support ES6 strict mode, `bar()` is undefined.  In ones that only support ES5 strict mode (e.g. IE10) it says: `In strict mode, function declarations cannot be nested inside a statement or block. They may only appear at the top level or directly inside a function body.`

**Comment by mhegazy** [comment]
i think  #7122 is a better one to track this issue,

**Comment by kitsonk** [comment]
Yes, I would agree #7122 is closer to being a dupe.  Thanks.
